### DISCUSSION TITLE
How to use Typescript with matched actions

### DISCUSSION THREAD

**Discussion by TacoV** [telegraf/telegraf#1974] [Q&A] [answered]
How can we set the correct Typescript type for the context, when using a regex to match a callback query (action)?

This works (but does not use regex)
```
import {Context} from "telegraf";
bot.action(
    'callback-value',
    (ctx:Context) => {
        console.log(
            "We got a callback from "+ctx.from?.first_name
        );
    }
);
```

This works (but uses 'any' type)
```
bot.action(
    /callback-(\d+)/,
    (ctx:any) => {
        console.log(
            "We got a callback with value " + ctx.match[1] +
            " from" + ctx.from?.first_name
        );
    }
);
```

But 'match' does not exist on Context. What should go in place of 'any'?

**Comment by MKRhere** [comment]
Remove the explicit type from ctx. Typescript will infer it.

**Comment by TacoV** [comment]
Did not know that - that works indeed!

In reality though I import a function from another file. I simplified the example in the question thinking it would work the same. This method won't transfer to that scenario, right?

The inferred type seems to be `Context<Update.CallbackQueryUpdate<CallbackQuery>> & Omit<Context<Update>, keyof Context<Update>> &`, I'm gonna try if something like that works.

**Reply by MKRhere** [reply, marked as answer]
You *can* do that, But the better way is to use Composers to split your bot.

```TS
// @file: part.ts

const composer = new Composer();
composer.action(
    'callback-value',
    ctx => {
        console.log(
            "We got a callback from "+ctx.from?.first_name
        );
    }
);

export default composer;
```

```TS
// @file: index.ts

import part from "./part.js";

const bot = new Telegraf();

// session, stage, etc goes here

// finally

bot.use(part);
```

This way you can split your bot into parts without having to do the hard work of typing it correctly yourself. If you have a custom context type, you can pass it to composer like `new Composer<MyCtx>()` as well.

**Reply by TacoV** [reply]
Thanks, that helps a lot.
